JsonPatch System.Text.Json - based .NET 10 Preview 4 update

wadepickett opened the pull request:
Fixes #35237
Fixes #35435

- Follows the detailed guidance provided in #35237 by @mkArtakMSFT.
- Add security mentions in the .NET 9 version of this topic.
- Mention the new System.Text.Json based version in the .NET 9 version of this topic and point to the new one to help alleviate confusion between the two.
- Update the app sample with the new code snippets and verify.
- Dynamic objects are not supported. Clarify that and remove dynamic object examples.
- Update any .NET/library API references with xref links

Table of changes with summaries:

| Change Type                     | Description                                                              | Line Numbers (approximate)     |
|----------------------------------|--------------------------------------------------------------------------|-------------------------------|
| Security warning (IMPORTANT)     | Added warning about security risks                                       | 5–7, 301–304                  |
| Mitigating security risks section| Added detailed attack/mitigation section                                 | 242–288, 534–580              |
| Remove dynamic object example    | Removed 'Dynamic objects' section and example code                       | 150–155 (in old orginal)                 |
| Remove dynamic-specific bullets  | Removed dynamic/static object distinctions under add, remove, move ops   | 161–162, 172–175, 195 (in old original)   |
| System.Text.Json clarification   | Clarified lack of support, proper formatter setup                        | 22–37 (edit)                  |
| Reorg/clarity of install steps   | Improved doc structure for package install/configuration                 | 7–13 (edit)                   |



<!-- PREVIEW-TABLE-START -->

---

#### Internal previews

| 📄 File | 🔗 Preview link |
|:--|:--|
| [aspnetcore/web-api/jsonpatch.md](https://github.com/dotnet/AspNetCore.Docs/blob/511a0b6daa6c7cf88036a00d98d58266fc8b1ae3/aspnetcore/web-api/jsonpatch.md) | [aspnetcore/web-api/jsonpatch](https://review.learn.microsoft.com/en-us/aspnet/core/web-api/jsonpatch?branch=pr-en-us-35571) |


<!-- PREVIEW-TABLE-END -->

wadepickett commented:
@mkArtakMSFT, could use your review.  Is there a different alias I should use that is recognized by github for internal review?  The one you used to create the issue does not seem to work.

wadepickett commented:
> My only concern was the pass on the .NET 10 What's New guidance when the community member opened an issue about the package reference ...
> 
> https://github.com/dotnet/AspNetCore.Docs/pull/35610/files
> 
> So far, nothing is broken here. 🤞🍀

I missed that occurred, thanks.  Ther was no issue tied to it and I wasn't on the loop.  I didn't see it in time on the PR list.  Looking it over now...

wadepickett commented:
I point to the correct one here:
" JSON Patch support with `System.Text.Json`, install the [`Microsoft.AspNetCore.JsonPatch.SystemTextJson`](https://www.nuget.org/packages/Microsoft.AspNetCore.JsonPatch/) NuGet package."

However, that is with the installation instructions.  I'll mention it at the very top too as we did before, but of course to the new package.

mikekistler commented:
Here's a sample project that illustrates the new JSON Patch support in both a controller-based and Minimal APIs app (actually it is one app with both controllers and minimal API endpoints).

https://github.com/mikekistler/aspnet-json-patch/tree/main

mikekistler commented on aspnetcore/web-api/jsonpatch.md:
````diff
@@ -14,226 +14,436 @@ uid: web-api/jsonpatch
 ```
 
-The changes made by applying a JSON Patch document to a resource are atomic. If any operation in the list fails, no operation in the list is applied.
+## JSON Patch Operations
````
I recommend we drop this whole section and just provide the examples.

mikekistler reviewed (changes requested):
I think we also want at least a brief description of JSON Patch support in the Minimal APIs section of the doc.

wadepickett commented:
> Here's a sample project that illustrates the new JSON Patch support in both a controller-based and Minimal APIs app (actually it is one app with both controllers and minimal API endpoints).
> 
> https://github.com/mikekistler/aspnet-json-patch/tree/main
Excellent, thanks @mikekistler! 
Can I put a copy of your example app here with the content or do you mean to just use it as a general guideline?  Or is the sample moving an MS owned repo and we link to it?

wadepickett commented on aspnetcore/web-api/jsonpatch.md:
```diff
@@ -14,226 +14,436 @@ uid: web-api/jsonpatch
+The sample action method:
 
-In the preceding JSON:
+* Constructs a `Customer`.
```
Removed entire operations section as recommended, so resolved for this section.  However, for the other examples we will use a db.

tdykstra commented on aspnetcore/web-api/jsonpatch.md:
```diff
@@ -1,11 +1,11 @@
 ---
 title: JsonPatch in ASP.NET Core web API
```
Looks like you have Json Patch as two words 78 times and one word 93 times; we should probably standardize on one word

wadepickett replied:
I will recheck, thanks.  Both are needed but I will verify I have t hem correct.

"JSON Patch" refers to the standard/format as defined in [RFC 6902](https://tools.ietf.org/html/rfc6902).
Example:

"JSON Patch is a standard format for describing changes to apply to a JSON document."
"The JSON Patch standard has inherent security risks."
Why:
This is the official name of the specification and should always be written as "JSON Patch" when discussing the general concept, the protocol, or the standard itself.

"JsonPatch" (no space, PascalCase)
Correct Usage:
"JsonPatch" is used as a proper noun for .NET types, classes, or package names.
Examples include:

Microsoft.AspNetCore.JsonPatch
JsonPatchDocument
JsonPatchError
Microsoft.AspNetCore.JsonPatch.SystemTextJson (NuGet package)

wadepickett replied:
Follow up:
The two different usages seem to be correct in the topic for each case:

"JSON Patch" is used for the standard/concept.
"JsonPatch" is used for .NET-specific code, classes, or package names.

tdykstra commented on aspnetcore/web-api/jsonpatch.md:
```diff
@@ -14,119 +14,54 @@ uid: web-api/jsonpatch
+JSON Patch support in ASP.NET Core web API is based on <xref:System.Text.Json> serialization, and requires the [`Microsoft.AspNetCore.JsonPatch.SystemTextJson`](https://www.nuget.org/packages/Microsoft.AspNetCore.JsonPatch.SystemTextJson) NuGet package.
 
-JSON Patch support in ASP.NET Core web API is based on `Newtonsoft.Json` and requires the [`Microsoft.AspNetCore.Mvc.NewtonsoftJson`](https://www.nuget.org/packages/Microsoft.AspNetCore.Mvc.NewtonsoftJson/) NuGet package. To enable JSON Patch support:
+**[JSON Patch](https://jsonpatch.com/)**:
```
What is jsonpatch.com?

wadepickett replied:
jsonpatch.com is a website dedicated to the JSON Patch standard.  It is not helpful for me to just put the link in the name here though.  I will move the link out, put it in the description below and indicate to go there for an overview of JSON Patch and the RFC 6902 standard so it has proper context.  Thanks.

wadepickett replied:
Resolved.  Moved it and added context.

tdykstra commented on aspnetcore/web-api/jsonpatch.md:
```diff
@@ -159,81 +95,211 @@ The preceding action method example calls an overload of `ApplyTo` that takes mo
+> ***This is not an exhaustive list of threats.*** App developers must conduct their own threat model reviews to determine an app-specific comprehensive list and come up with appropriate mitigations as needed. For example, apps which expose collections to patch operations should consider the potential for algorithmic complexity attacks if those operations insert or remove elements at the beginnin ...
 
-## The copy operation
+By running comprehensive threat models for their own apps and addressing identified threats while following the recommended mitigations below, consumers of these packages can integrate JSON Patch functionality into their apps while minimizing security risks.
```
Very long sentence, And most of it seems to be repeated in the following text

wadepickett replied:
Resolved.  Yes, that really needed to be whittled down!

tdykstra commented on aspnetcore/web-api/jsonpatch/includes/jsonpatch9.md:
```diff
@@ -2,9 +2,14 @@
 This article explains how to handle JSON Patch requests in an ASP.NET Core web API.
 
+> [!IMPORTANT]
+> The JSON Patch standard has ***inherent security risks***. Since these risks are inherent to the JSON Patch standard, this implementation ***doesn't attempt to mitigate inherent security risks***. It's the responsibility of the developer to ensure that the JSON Patch document is safe to apply to the target object. For more information, see the [Mitigating Security Risks](#mitigating-security-ri ...
```
```suggestion
> The JSON Patch standard has ***inherent security risks***. This implementation ***doesn't attempt to mitigate these inherent security risks***. It's the responsibility of the developer to ensure that the JSON Patch document is safe to apply to the target object. For more information, see the [Mitigating Security Risks](#mitigating-security-risks) section.
```

wadepickett replied:
I don't think I can apply this suggestion.  This was worked out with cela and the PM.  I think they deliberately put it in that order.

tdykstra reviewed (commented):
Nice work! I left a few comments and suggestions.

wadepickett commented on aspnetcore/web-api/jsonpatch.md:
```diff
@@ -14,226 +14,436 @@ uid: web-api/jsonpatch
+* Provides improved performance and reduced memory usage compared to the legacy `Newtonsoft.Json`-based implementation. For more information on the legacy `Newtonsoft.Json`-based implementation, see the [.NET 9 version of this article](xref:web-api/jsonpatch?view=aspnetcore-9.0&preserve-view=true).
 
-JsonPatch requires setting the `Content-Type` header to `application/json-patch+json`.
+The following benchmarks compare the performance of the new `System.Text.Json` implementation with the legacy `Newtonsoft.Json` implementation:
```
Removed.  Good point, thanks.

wadepickett commented on aspnetcore/web-api/jsonpatch.md:
```diff
@@ -14,226 +14,436 @@ uid: web-api/jsonpatch
+|                            | System.Text.JsonPatch  | 7.917 µs   | 7 KB             |
 
-The `System.Text.Json`-based input formatter doesn't support JSON Patch. To add support for JSON Patch using `Newtonsoft.Json`, while leaving the other input and output formatters unchanged:
+These benchmarks highlight significant performance gains and reduced memory usage with the new implementation.
```
Resolved.  Removed any indication of "new" implementation.  Thanks.

wadepickett commented on aspnetcore/web-api/jsonpatch.md:
```diff
@@ -14,226 +14,436 @@ uid: web-api/jsonpatch
-### Resource example
+:::code language="csharp" source="~/web-api/jsonpatch/samples/3.x/api/Controllers/HomeController.cs" id="snippet_PatchAction" highlight="1,3,9":::
+
+This code from the sample app works with the following `Customer` model:
```
Resolved.  Good catch, thanks!

wadepickett commented on aspnetcore/web-api/jsonpatch.md:
```diff
@@ -14,226 +14,436 @@ uid: web-api/jsonpatch
+## JSON Patch Operations
 
-## Path syntax
+The following sections describe the supported JSON Patch operations `add`, `remove`, `replace`, `move`, `copy`, and `test` that modify a JSON document and provide examples of their usage.
```
This section was removed.  However, I did covert to a list as you suggest here, for something similar where we describe the JSON Patch standard.  Thanks.

wadepickett commented on aspnetcore/web-api/jsonpatch.md:
```diff
@@ -14,226 +14,436 @@ uid: web-api/jsonpatch
-:::code language="csharp" source="~/web-api/jsonpatch/samples/6.x/api/Models/Customer.cs":::
+* If `path` points to an array element**: Copies from the element at `from` to the location specified by `path`, then removes the element at `from`.
+* If `path` points to a property**: Copies the value of the `from` property to the `path` property, then removes the `from` property.
+* If `path` points to a nonexistent property**:
```
Resolved as this section on operations was removed.

wadepickett commented on aspnetcore/web-api/jsonpatch.md:
```diff
@@ -14,226 +14,436 @@ uid: web-api/jsonpatch
-Zero-based indexes are used to specify array elements. The first element of the `addresses` array would be at `/addresses/0`. To `add` to the end of an array, use a hyphen (`-`) rather than an index number: `/addresses/-`.
+* If `path` points to an array element: Inserts a new element before the one specified by `path`.
+* If `path` points to a property: Sets the property value.
+* If `path` points to a nonexistent location:
```
Resolved as this section on operations was removed.

wadepickett commented:
Set back to draft.  
Added mikes app sample to samples/10.x/
Updated controller action method section to address the new code examples.

Next:
Add Minimal API equivalent to action method section for Min API.
Updating section: Apply a JSON Patch document to an object to use the newer app sample.

mikekistler reviewed (approved):
I think there is more we want to do to document JsonPatch, like adding something into the Minimal APIs section of the docs, but what's here seems complete and accurate for controllers (modulo the one comment I left on fixing the refs) and I don't think we should wait to merge this, so approving.

wadepickett commented:
Follow up:  I verified we are only referring to the new JsonPatch.SystemTextJson.JsonPatchDocument API for any **.NET 10** content in this repo.  Such as replacing:
<xref:Microsoft.AspNetCore.JsonPatch.JsonPatchDocument%601>
<xref:Microsoft.AspNetCore.JsonPatch.JsonPatchDocument.ApplyTo(System.Object)>

with:
<xref:Microsoft.AspNetCore.JsonPatch.SystemTextJson.JsonPatchDocument%601>
<xref:Microsoft.AspNetCore.JsonPatch.SystemTextJson.JsonPatchDocument.ApplyTo(System.Object)>
